src/java.base/share/classes/jdk/internal/loader/URLClassPath.java

```diff
@@ -439,34 +438,56 @@ private synchronized Loader getLoader(int index) {
                 continue;
             }
             // Otherwise, create a new Loader for the URL.
-            Loader loader;
+            Loader loader = null;
+            final URL[] loaderClassPathURLs;
             try {
                 loader = getLoader(url);
                 // If the loader defines a local class path then add the
                 // URLs as the next URLs to be opened.
-                URL[] urls = loader.getClassPath();
-                if (urls != null) {
-                    push(urls);
-                }
+                loaderClassPathURLs = loader.getClassPath();
             } catch (IOException e) {
-                // Silently ignore for now...
+                // log the error and close the unusable loader (if any)
+                if (DEBUG) {
+                    System.err.println("Failed to construct a loader or construct its" +
+                            " local classpath for " + url + ", cause:" + e);
+                }
+                if (loader != null) {
+                    closeQuietly(loader);
+                }
                 continue;
             } catch (SecurityException se) {
-                // Always silently ignore. The context, if there is one, that
-                // this URLClassPath was given during construction will never
-                // have permission to access the URL.
+                // log the error and close the unusable loader (if any).
+                // The context, if there is one, that this URLClassPath was
+                // given during construction will never have permission to access the URL.
                 if (DEBUG) {
                     System.err.println("Failed to access " + url + ", " + se );
                 }
+                if (loader != null) {
+                    closeQuietly(loader);
+                }
                 continue;
             }
+            if (loaderClassPathURLs != null) {
```

Comment:
I'm missing something here. Why does the compiler not complain that `loaderClassPathURLs` might not be initialized, when it's accessed here?

Reply:
Because we would not reach here - since it's either assigned at line 447 or we continue at line 457, 468, or exit throwing an uncaught exception?

Reply:
Is the if statement at line 470 redundant then?

Reply:
> Is the if statement at line 470 redundant then?

Reply:
What if `getLoader(url)` throws an exception so `Loader.getClassPath()` is not called?